ufl/algorithms/check_arities.py

```diff
@@ -43,6 +43,10 @@ def nonlinear_operator(self, o):
 
     expr = nonlinear_operator
 
+    def external_operator(self, o):
+        res = tuple(map_expr_dag(self, arg, compress=False)[0] for arg, _ in o.arguments())
```

Comment:
Can just do `res = tuple(map(self, (a for a, _ in o.arguments()))` I think. There's no need to invoke the map_expr_dag machinery because the arguments are terminal objects so there's an immediate non-recursive handler for them.

Reply:
Actually `arg` here is not necessarily a `ufl.Argument` it can be an expression containing an argument as explained in the class definition. Let’s consider the Gateaux derivative `dp’(u)[v] = dp * v` where `p` and `dp` are ExternalOperators and `dp` has a derivatives multi-index equal to (1,). Currently in the symbolic where the ExternalOperator is global and you want its action you don’t return  `dpdu * v` but `dp[(u,); v]` where `dpdu` is an external operator such that `dpdu.arguments() = (v,)`. 

But now if we consider dp’(cos(u))[v]  = dp * -sin(u) * v, we end up with `dp[(cos(u),); -sin(u)*v]` and dp.arguments() = (-sin(u)*v,) , so we need to hit the product handler.

What is a bit misleading here is that `arguments` refers to the expression on which we apply the action and not exclusively to the argument contained in this expression.